### PR TITLE
[DO NOT MERGE] Performance stuffs

### PR DESCRIPTION
**Pull request by spike01**
* Adds profiling code, you can change the profile's filename by changing 
https://github.com/thewhodidthis/gtx/blob/28573dcc65585a7e6cd184df24d5d560f3800dae/main.go#L41
(I'd normally put this behind a flag but I was lazy)

* Once you have a profile, you can look at it with 
```
go tool pprof -http=localhost:6060 gtx_buf_writer_parallel_branch.prof
```

* I noticed initially that most of the time was spent in `syscall.syscall` so used `dtruss` (the `strace` equivalent on OSX) to have a look at what's happening. Turns out calling `Write` on the file handle returned directly from `os.Create` writes line-by-line, which resulted in A LOT of syscalls to `write(2)`:

https://github.com/thewhodidthis/gtx/blob/28573dcc65585a7e6cd184df24d5d560f3800dae/performance.md?plain=1#L205

Wrapping those calls in a buffered writer reduced the number of calls significantly:

https://github.com/thewhodidthis/gtx/blob/28573dcc65585a7e6cd184df24d5d560f3800dae/project.go#L367-L373

https://github.com/thewhodidthis/gtx/blob/28573dcc65585a7e6cd184df24d5d560f3800dae/performance.md?plain=1#L257

But also didn't really improve performance 🙃 I got distracted by something else before going in much deeper but it looks like the Go runtime holds a lock somewhere related to creating new file handles, which is probably where the bottleneck is - since every action creates and writes to a file, adding concurrency just bottlenecks harder.

I added in a small check from the original `git2html`, which speeds things up a bit by reducing redundant work:

https://github.com/thewhodidthis/gtx/blob/28573dcc65585a7e6cd184df24d5d560f3800dae/project.go#L116-L121

lifted from here:

https://github.com/Hypercubed/git2html/blob/b29cc9517980058cce3f550c4befcf73ee8147bc/git2html.sh#L373-L380

But, this wasn't very satisfying as there are some bits of this that _should_ be possible to speed up with goroutines. I might come back to this at some point but was a fun hour or so poking around :)

### CODE REVIEW

File: performance.md [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+      4      3 write(0x7, "22/5a628ea0029a8067b87edae8291c6b58eba387\0", 0x29)		 = 41 0
+      4      2 write(0x7, ".html#L\0", 0x7)		 = 7 0
+      5      3 write(0x7, "11\0", 0x2)		 = 2 0
+      4      2 write(0x7, "\" id=\"L\0", 0x7)		 = 7 0
+     35      4 __semwait_signal(0x1F03, 0x0, 0x1)		 = -1 Err#60
+      5      3 write(0x7, "11\0", 0x2)		 = 2 0
+      6      4 write(0x7, "\">\0", 0x2)		 = 2 0
+      4      3 write(0x7, " 11\0", 0x3)		 = 3 0
+      4      2 write(0x7, "</a><br>\0", 0x8)		 = 8 0
+      4      2 write(0x7, "<a href=\"object/\0", 0x10)		 = 16 0
```
Comment (spike01) [inline performance.md:129]: You can see tiny tiny chunks of html being written here, which was what suggested to me that writes weren't being buffered

**Comment by thewhodidthis** [comment]
> I added in a small check from the original `git2html`, which speeds things up a bit by reducing redundant work:
> 
> https://github.com/thewhodidthis/gtx/blob/28573dcc65585a7e6cd184df24d5d560f3800dae/project.go#L116-L121

Thanks for adding this in, it does seem to help a great deal. I mistakenly thought the `os.MkdirAll()` call just after automatically took care of that too for some foolish reason, don't know why 😊

**Comment by spike01** [comment]
Closing this as I've just started a new job and want to be at PR Inbox Zero. Was super fun working on this, hope all's good with you!

**Comment by thewhodidthis** [comment]
Cool thanks @spike01, likewise! x

**Pull request closed without merging**